File: day9.py
#day9.py

def sumin(pre, v):
    s = set(pre)
    for x in pre:
        if v - x in s and x+x != v:
            return True
    return False

def part1(xmas, preamble=25):
    return [xmas[i] for i in range(preamble, len(xmas)) if not sumin(xmas[i-preamble:i],xmas[i])][0]

def findsum(slice, v):
    s = 0
    for i in range(len(slice)):
        s += slice[i]
        if s == v:
            return slice[:i+1]
        elif s > v:
            return None
    return None

def part2(xmas, v):
    for i in range(len(xmas)):
        f = findsum(xmas[i:], v)
        if f:
            return min(f) + max(f)

if __name__ == '__main__':
    xmas = [int(x) for x in open('input9.txt', 'r')]
    p1 = part1(xmas)
    p2 = part2(xmas,p1)
    print(p1)
    print(p2)

File: day9.rs
//day9.rs
use std::fs;

fn sumin(prefix: &[i64], v: i64) -> Option<&i64> {
    prefix
        .iter()
        .filter(|x| prefix.contains(&(v - *x)) && (*x+*x) != v)
        .next()
}

fn part1(xmas: &Vec<i64>, preamble: usize) -> i64 {
    for i in preamble..(xmas.len()) {
        if sumin(&xmas[i-preamble..i], xmas[i]).is_none() {
            return xmas[i];
        }
    }
    0
}

fn findsum(slice: &[i64], v: i64) -> Option<&[i64]> {
    let mut s = 0;
    for i in 0..slice.len() {
        s += slice[i];
        if s == v {
            return Some(&slice[0..i+1]);
        }
        else if s > v {
            return None
        }
    }
    None
}

fn part2(xmas: &Vec<i64>, v: i64) -> i64 {
    for i in 0..xmas.len() {
        if let Some(f) = findsum(&xmas[i..], v) {
            return f.iter().min().unwrap() + f.iter().max().unwrap()
        }
    }
    0
}

fn main() {
    let xmas: Vec<i64> = 
        fs::read_to_string("input9.txt").unwrap()
        .split("\n")
        .filter_map(|v| v.parse::<i64>().ok())
        .collect();
    let p1 = part1(&xmas, 25);
    let p2 = part2(&xmas, p1);

    println!("{}", p1);
    println!("{}", p2);
}
